crates/sui-framework/packages/sui-framework/sources/k_of_n_upgrade_policy.move

```diff
@@ -0,0 +1,522 @@
+            if (address != signer) {
+                let ballot_uid = object::new(ctx);
+                let ballot_id = object::uid_to_inner(&ballot_uid);
+                transfer::transfer(
+                    Ballot {
+                        id: ballot_uid,
+                        owner: signer,
+                        transfers_count: 0,
+                        digest,
+                        proposed_upgrade: proposal_id,
+                    },
+                    address,
+                );
+                vec_set::insert(&mut allowed_voters, ballot_id);
+            }
+        };
+
+        event::emit(UpgradeProposed {
+            upgrade_cap: cap_id,
+            proposal: proposal_id,
+            digest,
+            proposer: signer,
+            voters: cap.voters,
+            ballots: allowed_voters,
+        });
+
+        transfer::share_object(ProposedUpgrade {
+            id: proposal_uid,
+            cap: cap_id,
+            signer,
+            digest,
+            allowed_voters,
+            current_voters: vec_set::empty(),
+        })
+    }
+
+    /// Vote in favor of an upgrade, aborts if the ballot is not for the proposed
+    /// upgrade or if the upgrade was already performed.
+    public fun vote(
+        proposal: &mut ProposedUpgrade, 
+        ballot: Ballot,
+        ctx: &TxContext,
+    ) {
+        assert!(proposal.signer != @0x0, EAlreadyIssued);
+        let Ballot { 
+            id, 
+            owner: _,
+            transfers_count: _,
+            digest,
+            proposed_upgrade,
+        } = ballot;
+        assert!(digest == proposal.digest, EInvalidBallot);
+        let proposal_id = object::id(proposal);
+        assert!(proposal_id == proposed_upgrade, EInvalidBallot);
+        let ballot_id = object::uid_to_inner(&id);
+        vec_set::insert(&mut proposal.current_voters, ballot_id);
+        object::delete(id);
+
+        event::emit(UpgradeVoted {
+            proposal: object::id(proposal),
+            digest,
+            ballot: ballot_id,
+            voter: tx_context::sender(ctx),
+        });
+    }
+
+    /// Issue an `UpgradeTicket` for the upgrade being voted on.  Aborts if 
+    /// there are not enough votes yet, or if the upgrade was already performed.
+    /// The signer of the transaction must be the same as the one proposing the
+    /// upgrade.
+    public fun authorize_upgrade(
+        cap: &mut KofNUpgradeCap,
+        proposal: &mut ProposedUpgrade, 
+        ctx: &TxContext,
+    ): UpgradeTicket {
+        assert!(proposal.cap == object::id(cap), EInvalidProposalForUpgrade);
+        assert!(
+            vec_set::size(&proposal.current_voters) >= cap.required_votes, 
+            ENotEnoughVotes,
+        );
+
+        // assert the signer is the proposer and the upgrade has not happened yet
+        let signer = tx_context::sender(ctx);
+        assert!(proposal.signer != @0x0, EAlreadyIssued);
```

Comment:
Hello, this is Todd adding _very_ helpful comments that are not at all just about features he added that people aren't using... but you can do `@0` if you wish (decimal or hex works with addresses)